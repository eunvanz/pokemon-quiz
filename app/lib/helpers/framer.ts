import { Variants } from 'framer-motion'

export const motionVariants: Variants = {
  vibe: {
    translateX: [0, -4, 0, 4, 0],
    transition: {
      repeat: 2,
      duration: 0.1,
    },
  },
}
