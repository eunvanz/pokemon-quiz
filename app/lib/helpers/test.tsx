import CommonProvider from '@/app/components/common-provider'
import { PropsWithChildren } from 'react'
import api from '../api'

export interface TestProviderProps {
  api?: Partial<typeof api>
}

export const TestProvider: React.FC<PropsWithChildren<TestProviderProps>> = ({
  children,
  api = {},
}) => <CommonProvider api={api}>{children}</CommonProvider>
