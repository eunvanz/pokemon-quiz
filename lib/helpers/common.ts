export const checkIsSSR = () => {
  return typeof window === 'undefined'
}
