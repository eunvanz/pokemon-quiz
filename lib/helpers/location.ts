import { countries, zones } from 'moment-timezone/data/meta/latest.json'

export const getTimeZoneToCountryMap = () => {
  const timeZoneToCountryMap: Record<string, { name: string; abbr: string }> =
    {}
  Object.keys(zones).forEach((zone) => {
    const cityArr = zone.split('/')
    const city = cityArr[cityArr.length - 1]
    // @ts-ignore
    timeZoneToCountryMap[city] = countries[zones[zone].countries[0]]
  })
  return timeZoneToCountryMap
}

export const getUserLocationFromTimeZone = () => {
  const timeZone = Intl?.DateTimeFormat().resolvedOptions().timeZone
  if (timeZone) {
    const tzArr = timeZone.split('/')
    const region = tzArr[0]
    const city = tzArr[tzArr.length - 1]
    const { name, abbr } = getTimeZoneToCountryMap()[city]
    return {
      city,
      country: name,
      countryCode: abbr,
      region,
    }
  } else {
    return undefined
  }
}

export const getCountryFlagImgUrl = (countryCode: string) => {
  return `http://purecatamphetamine.github.io/country-flag-icons/3x2/${countryCode.toUpperCase()}.svg`
}
