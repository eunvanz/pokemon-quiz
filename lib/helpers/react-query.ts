import { QueryClient } from 'react-query'
import { Pageable } from '../types'

export const queryClient = new QueryClient()

export const getMergedPageData = <T>(data: Pageable<T>[]) => {
  return data.reduce((prev: T[], item) => [...prev, ...item.items], [])
}
