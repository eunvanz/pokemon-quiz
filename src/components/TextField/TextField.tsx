import { DetailedHTMLProps, forwardRef, InputHTMLAttributes, useMemo } from "react";
import { motion, useAnimation, Variants } from "framer-motion";
import tw from "twin.macro";

const motionVariants: Variants = {
  vibe: {
    translateX: [0, -4, 0, 4, 0],
    transition: {
      repeat: 5,
      duration: 0.1,
    },
  },
};

export interface TextFieldProps
  extends DetailedHTMLProps<InputHTMLAttributes<HTMLInputElement>, HTMLInputElement> {
  isBlock?: boolean;
  hasError?: boolean;
  errorMessage?: string;
}

const TextField = forwardRef<HTMLInputElement, TextFieldProps>(
  ({ isBlock, hasError, errorMessage, ...props }, ref) => {
    const cssByIsBlock = useMemo(() => {
      if (isBlock) {
        return tw`block w-full`;
      }
      return undefined;
    }, [isBlock]);

    const cssByErrorMessage = useMemo(() => {
      if (hasError) {
        return tw`border-red-500!`;
      }
      return undefined;
    }, [hasError]);

    return (
      <motion.div
        css={tw`flex flex-col items-start`}
        animate={hasError ? "vibe" : undefined}
        variants={motionVariants}
      >
        <input
          ref={ref}
          css={[
            tw`
              p-3 border-2 border-solid rounded-xl border-gray-200 transition-all text-base outline-none
            `,
            {
              "&:hover": props.disabled ? undefined : tw`border-gray-400`,
            },
            {
              "&:focus-visible": tw`border-2 border-blue-500`,
            },
            {
              "&::placeholder": tw`text-gray-400`,
            },
            {
              "&[disabled]": tw`bg-gray-200 cursor-not-allowed`,
            },
            cssByIsBlock,
            cssByErrorMessage,
          ]}
          {...props}
        />
        <div
          css={[
            tw`transition-all text-red-500 text-left overflow-hidden text-sm pl-2`,
            hasError && errorMessage ? tw`h-5` : tw`h-0`,
          ]}
        >
          {errorMessage || ""}
        </div>
      </motion.div>
    );
  },
);

export default TextField;
