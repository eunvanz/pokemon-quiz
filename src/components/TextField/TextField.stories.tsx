import type { ComponentStory, ComponentMeta } from "@storybook/react";
import { createStoryComponent } from "~/helpers/storybookHelpers";
import TextField from "./TextField";

export default {
  title: "components/TextField",
  component: TextField,
  args: {
    placeholder: "Placeholder",
  },
} as ComponentMeta<typeof TextField>;

const Template: ComponentStory<typeof TextField> = (args) => <TextField {...args} />;

export const Default = createStoryComponent(Template);

export const Disabled = createStoryComponent(Template, {
  disabled: true,
});

export const Error = createStoryComponent(Template, {
  hasError: true,
  errorMessage: "It's wrong answer",
});
