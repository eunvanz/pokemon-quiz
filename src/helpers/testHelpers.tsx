import { Story } from "@storybook/react";
import { render, RenderOptions } from "@testing-library/react";
import { QueryClientProvider } from "react-query";
import api from "~/api";
import { ApiProvider } from "~/api/apiContext";
import { queryClient } from "./reactQuery";

/**
 * render storybook component in test code
 * @param StoryComponent
 * @param props
 * @returns
 */
export const renderStory = <T,>(
  StoryComponent: Story<Partial<T>>,
  props?: Partial<T>,
  renderOptions?: RenderOptions,
) => {
  return render(<StoryComponent {...props} />, renderOptions);
};

export interface TestProviderProps {
  api: Partial<typeof api>;
}

export const TestProvider: React.FC<TestProviderProps> = ({ children, api }) => (
  <ApiProvider api={api}>
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  </ApiProvider>
);
