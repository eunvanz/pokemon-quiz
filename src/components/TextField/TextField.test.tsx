import { composeStories } from "@storybook/testing-react";
import { renderStory } from "~/helpers/testHelpers";
import { TextFieldProps } from "./TextField";
import * as stories from "./TextField.stories";

const { Default, Disabled, Error } = composeStories(stories);

describe("TextField", () => {
  describe("Default", () => {
    const setup = (props?: Partial<TextFieldProps>) => renderStory(Default, props);

    it("matches with previous snapshot.", () => {
      const { container } = setup();
      expect(container).toMatchSnapshot();
    });
  });

  describe("Disabled", () => {
    const setup = (props?: Partial<TextFieldProps>) => renderStory(Disabled, props);

    it("matches with previous snapshot.", () => {
      const { container } = setup();
      expect(container).toMatchSnapshot();
    });
  });

  describe("Error", () => {
    const setup = (props?: Partial<TextFieldProps>) => renderStory(Error, props);

    it("matches with previous snapshot.", () => {
      const { container } = setup();
      expect(container).toMatchSnapshot();
    });
  });
});
