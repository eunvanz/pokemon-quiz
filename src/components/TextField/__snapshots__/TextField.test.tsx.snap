// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField Default matches with previous snapshot. 1`] = `
<div>
  <div
    css="[object Object]"
  >
    <input
      css="[object Object],[object Object],[object Object],[object Object],[object Object],,"
      placeholder="Placeholder"
    />
    <caption
      css="[object Object],[object Object]"
    />
  </div>
</div>
`;

exports[`TextField Disabled matches with previous snapshot. 1`] = `
<div>
  <div
    css="[object Object]"
  >
    <input
      css="[object Object],[object Object],[object Object],[object Object],[object Object],,"
      disabled=""
      placeholder="Placeholder"
    />
    <caption
      css="[object Object],[object Object]"
    />
  </div>
</div>
`;

exports[`TextField Error matches with previous snapshot. 1`] = `
<div>
  <div
    css="[object Object]"
  >
    <input
      css="[object Object],[object Object],[object Object],[object Object],[object Object],,[object Object]"
      placeholder="Placeholder"
    />
    <caption
      css="[object Object],[object Object]"
    >
      It's wrong answer
    </caption>
  </div>
</div>
`;
